### PR TITLE
feat:slack connector

### PR DESCRIPTION
**Pull request by khareyash05**
Slack Connector

### CODE REVIEW

File: packages/apps/slack/src/services/slack.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    this.logger_.info(`Starting oculation of Slack for ${org.id} organisation`);
+
+    // Get Slack OAuth for the organisation
+    const oauth = await this.oauthService_.retrieve({
+      id: org.id,
+      app_name: AppNameDefinitions.SLACK,
+    });
+
+    if (!oauth) {
+      this.logger_.error(`No Jira OAuth found for ${org.id} organisation`);
```
Comment (louismurerwa) [inline packages/apps/slack/src/services/slack.ts:48]: Change Jira to Slack!

File: packages/ocular/core-config.js [hunk 2]
```diff
@@ -15,7 +15,7 @@ switch (process.env.NODE_ENV) {
     ENV_FILE_NAME = ".env.dev";
     break;
   default:
-    ENV_FILE_NAME = ".env.local";
+    ENV_FILE_NAME = ".env.dev";
```
Comment (louismurerwa) [inline packages/ocular/core-config.js:18]: change back to env.local

File: packages/apps/slack/src/services/slack.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        Accept: "application/json",
+      },
+    };
+
+    let documents: IndexableDocument[] = [];
+
+    try {
+      const slackChannels = await this.fetchSlackChannels(config)
+
+      for (const channel of slackChannels) {
```
Comment (louismurerwa) [inline packages/apps/slack/src/services/slack.ts:64]: Line 64 and 65 looks good!!

On line 66: I am thinking of creating documents on the basis of threads and not individual methods.

For example: 
- Iterate through the conversations
   if conversation.thread has not been seen
      - find all messages in that thread.
      - Create an IndexableDocument ( tittle = name of first messge in thread, Sections = [Messages]
      - mark thread seen.

File: packages/apps/slack/src/services/slack.ts [hunk 4]
```diff
@@ -60,37 +61,31 @@ export default class SlackService extends TransactionBaseService {
...
+            id:conversation.ts, // conversation id
             organisationId:org.id,
-            title:conversation.text,
             source:AppNameDefinitions.SLACK,
-            updatedAt: new Date(Date.now()),
-            metadata:{sentByUser:conversation.user}
+            title:conversation.text, // the main message which lead to conversation
+            metadata:{channel_id:channel.id}, // passing channel id just for top down reference
+            sections:sections, // an array of messages in the specific conversation
+            updatedAt: new Date(Date.now())
```
Comment (Vivek-Lahole) [inline packages/apps/slack/src/services/slack.ts:81]: Add `type: DocType.TEXT` in Indexable Document

File: packages/apps/slack/src/services/slack.ts [hunk 5]
```diff
@@ -156,4 +151,14 @@ export default class SlackService extends TransactionBaseService {
       throw new Error("Failed to fetch channel conversation.");
     }
   }
+
+  async fetchThreadForConversation(channelID, tsID, config){
```
Comment (Vivek-Lahole) [inline packages/apps/slack/src/services/slack.ts:155]: Provide types to arguments `channelID : string`

File: packages/apps/slack/src/services/slack.ts (hunk 4 above)
Comment (khareyash05) [reply packages/apps/slack/src/services/slack.ts:81]: in metadata?

Comment (Vivek-Lahole) [reply packages/apps/slack/src/services/slack.ts:81]: just like a title , add a field as `type : type: DocType.TEXT` , for reference you can checkout the jira app in Main ocular Repo

**Review by louismurerwa — APPROVED** [review]

File: packages/apps/slack/src/services/slack.ts [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    };
+
+    let documents: IndexableDocument[] = [];
+
+    try {
+      const slackChannels = await this.fetchSlackChannels(config)
+      for (const channel of slackChannels) {
+        const conversations = await this.fetchChannelConversations(channel.id,config);
+        for(const conversation of conversations){
+          const thread = await this.fetchThreadForConversation(channel.id,conversation.ts,config)
```
Comment (ArnavBalyan) [inline packages/apps/slack/src/services/slack.ts:68]: This seems to be a bug, .ts does not exist on conversation. Similarly, it's non-existent for the references in this file. Consider changing to `conversation.id`

**Review by ArnavBalyan — COMMENTED** [review]
Indexing and fetching does not seem to be working, code needs minor changes & testing.

File: packages/apps/slack/src/services/slack.ts (hunk 6 above)
Comment (khareyash05) [reply packages/apps/slack/src/services/slack.ts:68]: Hey id will not work . Please refer to the following doc [https://api.slack.com/methods/conversations.history](https://api.slack.com/methods/conversations.history)

File: packages/apps/slack/src/services/slack.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+  async fetchChannelConversations(channelID:string ,config:Config){
+    try{
+      const conversationsEndpoint = `https://slack.com/api/conversations.history?channel=${channelID}`
+      const response = await axios.get(conversationsEndpoint,config)
+      const conversationsArray = response.data.messages || []
+      const conversations = conversationsArray.map((conversations)=>({
+        id: conversations.ts,
+        text:conversations.text,
+        user: conversations.user
+      }))
```
Comment (ArnavBalyan) [inline packages/apps/slack/src/services/slack.ts:150]: We are assigning `conversations.ts` to `id`, not to `ts`

File: packages/apps/slack/src/services/slack.ts (hunk 6 above)
Comment (ArnavBalyan) [reply packages/apps/slack/src/services/slack.ts:68]: Yes, refer to the other comment on this file, `ts` does not exist on the conversation object

File: packages/apps/slack/src/services/slack.ts (hunk 7 above)
Comment (khareyash05) [reply packages/apps/slack/src/services/slack.ts:150]: Oh right! Missed that!

Comment (khareyash05) [reply packages/apps/slack/src/services/slack.ts:150]: Could you please create a PR with the change? Thanks for finding the issue

**Pull request merged**